torch/nn/modules/transformer.py

```diff
@@ -280,22 +296,25 @@ def __init__(self, d_model, nhead, dim_feedforward=2048, dropout=0.1):
         self.dropout2 = Dropout(dropout)
         self.dropout3 = Dropout(dropout)
 
-    def forward(self, tgt, memory, tgt_mask=None, memory_mask=None):
+    def forward(self, tgt, memory, tgt_mask=None,
+                memory_mask=None, key_padding_mask=None):
         r"""Pass the inputs (and mask) through the decoder layer.
 
         Args:
             tgt: the sequence to the decoder layer (required).
             memory: the sequnce from the last layer of the encoder (required).
             tgt_mask: the mask for the tgt sequence (optional).
             memory_mask: the mask for the memory sequence (optional).
+            key_padding_mask: the mask for the src keys per batch (optional).
 
         Shape:
             see the docs in Transformer class.
         """
         tgt2 = self.self_attn(tgt, tgt, tgt, attn_mask=tgt_mask)[0]
         tgt = tgt + self.dropout1(tgt2)
         tgt = self.norm1(tgt)
-        tgt2 = self.multihead_attn(tgt, memory, memory, attn_mask=memory_mask)[0]
+        tgt2 = self.multihead_attn(tgt, memory, memory, attn_mask=memory_mask,
+                                   key_padding_mask=key_padding_mask)[0]
```

Comment:
I think we want to add the key_padding_mask to the self-attention layer.

Reply:
Not the memory layer? I would think people would want to mask the padding pieces on the input side when decoding and that would be the way to achieve it. I suppose you might want to mask padding on the target. How about I just add key_padding_mask to self attention layer and memory_key_padding_mask arg for memory layer. In Transformer I will add source_key_padding_mask, target_key_padding_mask, and memory_key_padding_mask. That will cover every base that we would need and be as flexible as possible.

Reply:
I agree this will cover every base enough.